tls/extensions/s2n_client_key_share.c

```diff
@@ -306,8 +304,137 @@ static int s2n_client_key_share_send(struct s2n_connection *conn, struct s2n_stu
     return S2N_SUCCESS;
 }
 
-static int s2n_client_key_share_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
-{
+static int s2n_client_key_share_parse_ecc_share(struct s2n_stuffer *key_share, const struct s2n_ecc_named_curve *curve,
+        struct s2n_ecc_evp_params *ecc_params) {
+    notnull_check(key_share);
+    notnull_check(curve);
+    notnull_check(ecc_params);
+
+    struct s2n_blob point_blob = { 0 };
+    GUARD(s2n_ecc_evp_read_params_point(key_share, curve->share_size, &point_blob));
+
+    /* Ignore curves with points we can't parse */
+    ecc_params->negotiated_curve = curve;
+    if (s2n_ecc_evp_parse_params_point(&point_blob, ecc_params) < 0) {
+        ecc_params->negotiated_curve = NULL;
+        GUARD(s2n_ecc_evp_params_free(ecc_params));
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_client_key_share_recv_ecc(struct s2n_connection *conn, struct s2n_stuffer *key_share,
+        uint16_t curve_iana_id, bool *match) {
+    notnull_check(conn);
+    notnull_check(key_share);
+    notnull_check(match);
+
+    const struct s2n_ecc_preferences *ecc_pref = NULL;
+    GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
+    notnull_check(ecc_pref);
+
+    ENSURE_POSIX(s2n_ecc_preferences_includes_curve(ecc_pref, curve_iana_id), S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+
+    const struct s2n_ecc_named_curve *curve = NULL;
+    struct s2n_ecc_evp_params *client_ecc_params = NULL;
+    for (size_t i = 0; i < ecc_pref->count; i++) {
+        if (curve_iana_id == ecc_pref->ecc_curves[i]->iana_id) {
+            curve = ecc_pref->ecc_curves[i];
+            client_ecc_params = &conn->secure.client_ecc_evp_params[i];
+            break;
+        }
+    }
+    
+    notnull_check(curve);
+    notnull_check(client_ecc_params);
+
+    /* Ignore curves that we've already received material for */
```

Comment:
When is this case possible?

Reply:
It's possible if some rogue client sends multiple key shares for the same group. We accept the first key share, and ignore any that come after it. This is consistent with what we do currently:

https://github.com/awslabs/s2n/blob/69be9c0c6aa139cc9c064ed4592908427b8af02b/tls/extensions/s2n_client_key_share.c#L359-L363